include/geolib/Box.h

```diff
@@ -21,14 +21,18 @@ class Box : public Shape {
 
     bool intersect(const Vector3& p) const;
 
-    Box getBoundingBox() const;
+    const Box getBoundingBox() const;
```

Comment:
`const` is unneeded limiting here, you're giving a fresh copy of the box to the caller, you don't need to restrict what the caller can do with its box.

Reply:
I am returning *this. Is this a copy?

Reply:
The caller writes something like `const Box mybox = shape.getBoundingBox();`. In C/C++, a variable represents space to store the data (unlike in Python or Java where a variable is always/often a reference to data elsewhere). The `mybox` value lives in data space of the caller and has room for all data of a box. Unless the returned box data contains pointers or references to data of the callee (it doesn't in this case afaik), it's a completely self-contained value.

This is exactly the same as with integers (C/C++ works exactly the same for every type of data):
``` C
int i = 3; // "our" integer (object)
int j; // 'foreign' integer (object)
j = i; // makes a fully self-contained copy of i, and puts it in j.
```

Your `*this` is data owned by the box (in fact, it is literally the box object itself). The copying is being done in the `operator=` method/function for the integers example, or likely (not sure about the precise semantics here) the copy constructor `Box::Box(const Box&)` for the box assignment, as `mybox` hasn't been initialized until the call returns. (You can't do `const Box b; b = shape.getBoundingBox();` as that violates the `const` property of `b`. You may be allowed to do (not sure about the semantics of `const` in this case) `Box b; b = shape.getBoundingBox();` which would likely make a copy using `Box::operator=` from an temporary `const Box` variable created by the compiler (but check the cpp reference if you really want to know).)

Restricting modification of a result value is mostly useful if you give away some part of 'your own' object data for inspection by the caller, but that data should not be modified (as it's also being used by the box object), and you want to avoid making a full copy of it. You'd write `const Box &getBoundingBox()` in such a case, with the effect that the caller now gets a reference to data owned by the called `Shape`.
The advantage here is that it avoids copying, the disadvantage is that the shape object is now responsible for keeping that referenced data valid until all callers don't need the information any more.